ceph/mds.go

```diff
@@ -183,3 +217,204 @@ func (m *MDSCollector) Collect(ch chan<- prometheus.Metric, version *Version) {
+		return
+	}
+
+	hc := &healthDetailCheck{}
+
+	err = json.Unmarshal(data, hc)
+	if err != nil {
+		m.logger.WithError(err).Error("failed unmarshalling health detail")
+		return
+	}
+
+	check, ok := hc.Checks["MDS_SLOW_REQUEST"]
+	if !ok {
+		// No slow requests! Yay!
+		return
+	}
+
+	for _, cc := range check.Detail {
+		mdsNameParts := strings.Split(cc.Message, "(")
+		if len(mdsNameParts) != 2 {
+			m.logger.WithError(
+				errors.New("incorrect part count"),
+			).WithFields(logrus.Fields{
+				"message": cc.Message,
+			}).Error("invalid mds slow request message found, check syntax")
+			continue
+		}
+
+		mdsName := mdsNameParts[0]
+
+		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+		defer cancel()
+
+		data, err := runMDSStatus(ctx, m.config, m.user, mdsName)
+		if err != nil {
+			m.logger.WithField("mds", mdsName).WithError(err).Error("failed getting status from mds")
+			return
+		}
+
+		mss := &mdsStatus{}
+
+		err = json.Unmarshal(data, mss)
+		if err != nil {
+			m.logger.WithField("mds", mdsName).WithError(err).Error("failed unmarshalling mds status")
+			return
+		}
+
+		ctx, cancel = context.WithTimeout(context.Background(), 1*time.Minute)
+		defer cancel()
+
+		data, err = runBlockedOpsCheck(ctx, m.config, m.user, mdsName)
+		if err != nil {
+			m.logger.WithField("mds", mdsName).WithError(err).Error("failed getting blocked ops from mds")
+			return
+		}
+
+		mso := &mdsSlowOp{}
+
+		err = json.Unmarshal(data, mso)
+		if err != nil {
+			m.logger.WithField("mds", mdsName).WithError(err).Error("failed unmarshalling mds blocked ops")
+			return
+		}
+
+		for _, op := range mso.Ops {
+			if op.TypeData.OpType == "client_request" {
+				fso, err := extractOpFromDescription(op.Description)
+				if err != nil {
+					m.logger.WithField("mds", mdsName).WithError(err).Error("failed parsing blocked ops description")
+					continue
+				}
+
+				select {
+				case m.ch <- prometheus.MustNewConstMetric(
+					m.MDSBlockedOps,
+					prometheus.CounterValue,
+					1,
+					mss.FsName,
+					mdsName,
+					mss.State,
+					op.TypeData.OpType,
+					fso,
+					op.TypeData.FlagPoint,
+				):
+				default:
+				}
+
+				continue
+			}
+
+			select {
+			case m.ch <- prometheus.MustNewConstMetric(
+				m.MDSBlockedOps,
+				prometheus.CounterValue,
+				1,
+				mss.FsName,
+				mdsName,
+				mss.State,
+				op.TypeData.OpType,
+				"",
+				op.TypeData.FlagPoint,
+			):
+			default:
+			}
+		}
+
+	}
+}
+
+// extractOpFromDescription is designed to extract the fs optype from a given slow/blocked
+// ops description.
+//
+// For e.g. given the description as follows:
+//
+//	"client_request(client.20001974182:344151 rmdir #0x10000000030/72a26231-ac24-4f69-9350-8ebc5444c9ea 2024-02-13T22:11:00.196767+0000 caller_uid=0, caller_gid=0{})"
+//
+// we should be able to extract the following fs optype out of it:
+//
+//	"rmdir"
+func extractOpFromDescription(desc string) (string, error) {
```

Comment:
extracting the client could be useful here as well for tracing issues; don't need it for metrics, but troubleshooting tooling

Reply:
Agreed! Something we should evaluate alongside exposing inode info as well (which I was planning to do later to avoid time-series explosion). I can expose it from the function for now with a note, and later on we can see when we want to expose it to the scraper.

Reply:
Added!